### PR TITLE
BulkExecutorPerfChange-1

### PR DESCRIPTION
**Pull request by xinlian12**
**Issue**:
The current bulk executor does not has as good as perf compared to 4.16.0.

**How to repro:**
<p style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; margin: 0in; font-family: Calibri; font-size: 11pt;">Ingesting 1000000 item (&lt;1KB) total with iterator:</p><div style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; direction: ltr;">

Version | SDK version |  
-- | -- | --
  | 4.16.0 | 8m6s
  | 4.23.0-beta.1 | 14m55s

</div>

<p style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; margin: 0in; font-family: Calibri; font-size: 11pt;">Ingesting 8794 item(122kb each) total with iterator:</p><div style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; direction: ltr;">

Version | SDK version |  
-- | -- | --
  | 4.16.0 | 2m52s
  | 4.23.0-beta.1 | 7m2s

</div><p style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; margin: 0in; font-family: Calibri; font-size: 11pt;"> </p>

<p style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; margin: 0in; font-family: Calibri; font-size: 11pt;">Ingesting 587 item(1.8MB each) total with iterator:</p><div style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; direction: ltr;">

Version | SDK version |  
-- | -- | --
  | 4.16.0 | 45s
  | 4.23.0-beta.1 | 2m29s

</div>

**Root cause:**
Compared to 4.16.0, there are two major diff:
1. in 4.16.0, the default max batch concurrency is 2, while in current master it is 1
2. in 4.16.0, the default max batch flush interval is 100ms, while in current master it is 1000

**Fix:**
This PR contains the change target for root cause max batch concurrency. Changes included:
1. Expose MaxMicroBatchConcurrency in CosmosBulkExecutionOptions.
    It is better to expose the config so customers can tune the value case by case. For example, in spark, it is probably always enough to use default value 1, while for some customer using Java SDK directly, if RU is not one of their concern, or they are ingesting large docs, it is reasonable to increase to 2 to improve the perf.

2. Also use total serialized batch size to decide whether to flush the batch.
The current flow is every batch size or flush interval a new batch is going to be flushed. While we are validating the max body size limit  for this batch, items can not be sent will be put back into the queue again. 
when customer trying to insert large items (more than 2000KB), it basically mean only 1 item can be in the batch, most of the items will be put back into the queue, which is unnecessary work if we have the size guard.

3. Cache cosmos item operation serialized length.
There are many cases can cause an item be retried several times, exceeds max body size limit, 410s, 429s. Each time the item being back to the group sink, the size of the serialized format will need to be calculated again, which is not really necessary. We cache the calculated result for future use.

**Local test result:**
RU config 15000
<p style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; margin: 0in; font-family: Calibri; font-size: 11pt;">Ingesting 1000 items (1.8M each)</p><div style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; direction: ltr;">

Version | SDK version | Total time(ms)
-- | -- | --
  | PR | 171,159
  | Master | 201471
  | PR  + concurrency 2 | 93283

</div>

<p style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; margin: 0in; font-family: Calibri; font-size: 11pt;">Ingesting 10000 items (122KB each):<span class="Apple-converted-space"> </span></p><div style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; direction: ltr;">

Version | SDK version | Total time(ms)
-- | -- | --
  | PR | 221655
  | Master | 287093
  | PR  + concurrency 2 | 99323

</div>

<p style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; margin: 0in; font-family: Calibri; font-size: 11pt;">Ingesting 10,000 items (61KB each):</p><div style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; direction: ltr;">

Version | SDK version | Total time(ms)
-- | -- | --
  | PR | 114361
  | Master | 115440
  | PR  + concurrency 2 | 61233

</div>

<p style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; margin: 0in; font-family: Calibri; font-size: 11pt;">Ingesting 10,000 items (&lt;1KB each):<span class="Apple-converted-space"> </span></p><div style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; direction: ltr;">

Version | SDK version | Total time(ms)
-- | -- | --
  | PR | 6470
  | Master | 6649
  | PR  + concurrency 2 | 5215

</div>

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosBulkExecutionOptions.java [hunk 1]
```diff
@@ -92,13 +94,35 @@ CosmosBulkExecutionOptions setMaxMicroBatchSize(int maxMicroBatchSize) {
...
+     *
+     * Attention! Please adjust this value with caution.
+     * By increasing this value, more concurrent requests will be allowed to be sent to the server,
+     * in which case may cause 429 or request timed out due to saturate local resources, which could degrade the performance.
+     *
+     * @param maxMicroBatchConcurrency the micro batch concurrency.
+     *
+     * @return the bulk processing options.
+     */
+    public CosmosBulkExecutionOptions setMaxMicroBatchConcurrency(int maxMicroBatchConcurrency) {
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosBulkExecutionOptions.java:118]: should we have beta tag for these? or are these APIs stable enough?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/batch/CosmosItemOperationBase.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+import com.azure.cosmos.implementation.JsonSerializable;
+import com.azure.cosmos.models.CosmosItemOperation;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+public abstract class CosmosItemOperationBase implements CosmosItemOperation {
+
+    private final AtomicReference<Integer> serializedLengthReference;
+    private final AtomicReference<JsonSerializable> serializedOperation;
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/batch/CosmosItemOperationBase.java:14]: nice pattern. :-)

**Review by moderakh — APPROVED** [review]
LGTM. Thanks Annie. Awesome work.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosBulkExecutionOptions.java (hunk 1 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosBulkExecutionOptions.java:118]: IMO Stable enough - Annie has restricted the values so customers can't shoot themselves. And lifting the constraints would not be breaking. Naming is hopefully something we wouldn't have to churn on.

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM - Thanks!

**Comment by xinlian12** [comment]
Two tests are failing: 
conflictNonExistingCustomSproc
splitQueryContinuationToken

**Pull request merged**